Feature/reduce container creation delay

carpawell opened the pull request:
Closes #519.

Implement notary requests subscription and handling mechanisms that can be also
reused for other `node` -> `contract` ---`*notification*`---> `alphabet` -> `contract`
flows in the future.

Fix errors/typos in commentaries. Delete spaces in logs' keys.

alexvanin commented on cmd/neofs-node/container.go:
```diff
@@ -41,7 +41,7 @@ const (
 
 func initContainerService(c *cfg) {
-	wrap, err := wrapper.NewFromMorph(c.cfgMorph.client, c.cfgContainer.scriptHash, 0)
+	wrap, err := wrapper.NewFromMorph(c.cfgMorph.client, c.cfgContainer.scriptHash, 0, wrapper.TryNotary())
```
Check if it breaks container.Delete and container.SetEACL

carpawell replied:
yes, it breaks

alexvanin commented on pkg/morph/event/listener.go:
```diff
@@ -285,88 +349,236 @@ func (s listener) parseAndHandle(notifyEvent *state.NotificationEvent) {
+
+	// handle the event
+	l.mtx.RLock()
+	handler, ok := l.notaryHandlers[notaryKey]
```
Okay, notaryKey are quite different for notary case and notification case (they are different structures), so we can't unify it straight away.

carpawell commented on pkg/morph/client/notary.go:
```diff
@@ -282,10 +302,10 @@ func (c *Client) notaryInvokeAsCommittee(method string, args ...interface{}) err
 }
 
-func (c *Client) notaryInvoke(committee bool, contract util.Uint160, method string, args ...interface{}) error {
+func (c *Client) notaryInvoke(committee bool, invokedByAlpha bool, contract util.Uint160, method string, args ...interface{}) error {
```
fixed

carpawell commented on pkg/morph/client/notary.go:
```diff
@@ -479,15 +520,33 @@ func (c *Client) notaryWitnesses(multiaddr *wallet.Account, tx *transaction.Tran
 }
 
-func (c *Client) notaryMultisigAccount(ir []*keys.PublicKey, committee bool) (*wallet.Account, error) {
+func (c *Client) notaryMultisigAccount(ir []*keys.PublicKey, committee bool, invokedByAlpha bool) (*wallet.Account, error) {
```
fixed

carpawell commented on pkg/morph/event/notary.go:
```diff
@@ -0,0 +1,36 @@
+}
+
+// NotaryTypeFromBytes converts bytes slice to NotaryType.
+func NotaryTypeFromBytes(data []byte) NotaryType {
```
did just the same as in `notification.go`(not used too). for future "potential" usage I guess

delete?

carpawell commented on pkg/morph/client/constructor.go:
```diff
@@ -145,6 +145,25 @@ func WithSigner(signer *transaction.Signer) Option {
+// key.
+//
+// Considered to be used by IR nodes only.
+func IsAlphabet() StaticClientOption {
```
moved to `static.go`, also moved back `TryNotary` transferred by mistake

carpawell commented on cmd/neofs-node/config/morph/config.go:
```diff
@@ -4,14 +4,22 @@ import (
 	DialTimeoutDefault = 5 * time.Second
+
+	// NotaryDepositAmountDefault is a default deposit amount to notary contract.
+	NotaryDepositAmountDefault = 1_0000_0000 // 1.0 Fixed8
```
changed to 0.5

carpawell commented on pkg/innerring/bindings.go:
```diff
@@ -8,33 +8,53 @@ type (
+	// and NotaryContractProcessor interfaces.
+	NotaryContractProcessor interface {
+		ListenerNotaryParsers() []event.NotaryParserInfo
+		ListenerNotaryHandlers() []event.NotaryHandlerInfo
```
think it can be done after addition notary flow to all contract processors

carpawell commented:
@alexvanin, added resigning SN's notary request by IR instead of creating a new one

key changes: 
- `notaryPreparator` is struct now, not func. checking received requests is stricter now
- added resigning notary request in `morph/client.notary.go`
- added retrieving low-level morph client via `Morph` in wrappers
- `Put` event contains a raw notary request pointer. `Put` handling logic in IR processor depends on that fact

carpawell commented:
@AnnaShaleva, can you, please, take a look at https://github.com/nspcc-dev/neofs-node/pull/770/commits/ce63109cbd31e1a452a1ba61016ecb1abae1d54d and give us feedback about SN's notary request checking by alphabet node?

AnnaShaleva commented on pkg/morph/event/notary_preparator.go:
```diff
@@ -0,0 +1,243 @@
+	}
+
+	// retrieve contract's script hash
+	contractHash, err := util.Uint160DecodeBytesBE(ops[opsLen-2].param)
```
Don't we need to verify contract hash and method? Because now you allow *any* method of *any* contract to be called with *any* set of arguments which can be constructed from `PUSH*` opcodes. And this transaction will be successfully signed by alphabet nodes. It seemed to be quite dangerous.

AnnaShaleva commented on pkg/morph/client/notary.go:
```diff
@@ -296,6 +296,51 @@ func (c *Client) NotaryInvokeNotAlpha(contract util.Uint160, fee fixedn.Fixed8,
+		[]byte{byte(opcode.RET)},
+		-1,
+		0,
+		c.notary.fallbackTime,
```
To provide the same `NotValidBefore` value for all alphabet fallbacks, you can use `NotValidBefore` value of the original (i.e. sent by storage node) fallback transaction. Otherwise current `NotValidBefore` alphabet fallback values may slightly differ.

carpawell replied:
do you mean to pass `fallbackTx.ValidUntilBlock - fallbackTx.Attributes[NotValidBeforeT] + 1` as 5th arg here?

carpawell replied:
(`fallbackTx.Attributes[NotValidBeforeT]` is shorthand)

AnnaShaleva replied:
I just realised, that it is already the proper one, because `c.notary.fallbackTime` is a constant and fallback's `ValidUntilBlock` equals to the main's one, so it's OK to keep the existing code.

AnnaShaleva commented on pkg/morph/subscriber/subscriber.go:
```diff
@@ -143,6 +155,14 @@ func (s *subscriber) routeNotifications(ctx context.Context) {
 
 				s.blockChan <- b
+			case response.NotaryRequestEventID:
+				notaryRequest, ok := notification.Value.(*response.NotaryRequestEvent)
```
I see that removed requests are being filtered later, but anyway, do we need them at all?

roman-khimov replied:
To be fair that's the question I'm asking myself in general (subscription API-wise), who needs them?

carpawell replied:
just wanted to make base functionality in node listener: it can listen to, parse and handle any notary requests

but yes, there are no parsers and handlers for such events in the alphabet node now

/cc @alexvanin

AnnaShaleva replied:
I think it still can be a nice indicator of the case when neither main nor fallback transaction was completed. Other cases are covered by existing subscriptions API, e.g. subscriptions for transactions execution (main or fallback).

carpawell commented on pkg/morph/event/listener.go:
```diff
@@ -284,17 +349,72 @@ func (l listener) parseAndHandleNotification(notifyEvent *state.NotificationEven
+	}
+
+	// parse the notary event
+	event, err := parser(notaryEvent.Params())
```
there is a special method in the listener that registers parsers and handlers. this line does now execute for removed notary requests since there is no code that registers parsers and listeners for it

carpawell commented on pkg/morph/event/notary_preparator.go:
```diff
@@ -0,0 +1,243 @@
+
+func (p Preparator) validateWitnesses(w []transaction.Witness, alphaKeys keys.PublicKeys) error {
+	// notary request's main tx is expected to have
+	// at least three witnesses: one for proxy contract,
```
in fact, it is going to be used in other notification methods, not only `Put`but yes, I think it can not differs from 3

carpawell commented on pkg/morph/event/notary_preparator.go:
```diff
@@ -0,0 +1,243 @@
+	}
+
+	// validate main TX's witnesses
+	err = p.validateWitnesses(nr.MainTransaction.Scripts, currentAlphabet)
```
1. such checks as: exactly 3 signers; second is multiacc of Alphabet? the first one is always proxy(does not need to check, I guess, since it is filtered in subscription process), and the third one is always notary(does not need to check, I guess, since this check is done with notary service), am I correct?
2. agree, will add such check

AnnaShaleva replied:
Yes, that's true.

alexvanin reviewed (approved):
Let's rename `IsAlphabet` to `AsAlphabet` and call it a day for this PR.
In next PR I expect to see all container related notification with notary notification support, merged interfaces of `NotaryContractProcessor` and `ContractProcessor` and small processor event handling refactor.

AnnaShaleva commented on pkg/morph/event/notary_preparator.go:
```diff
@@ -0,0 +1,316 @@
+}
+
+func (p Preparator) validateExpiration(fbTX *transaction.Transaction) error {
+	if len(fbTX.Attributes) != 3 || fbTX.Attributes[1].Type != transaction.NotValidBeforeT {
```
The attributes order is undefined in general, so here you check the way neo-go RPC client constructs them. So I suggest to keep attributes length check and use `fbTX.GetAttributes(transation.NotValidBefore)` instead of fetching attribute by index.

carpawell replied:
thanks, added more general attribute check
